### PR TITLE
Modifying IGP to be a hook

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

- IGP as a standalone hook, implementing postDispatch to call payForGas directly
- Setting a DEFAULT_GAS_USAGE if metadata not specified and message.senderAddress() as refund address if not specified.

### Drive-by changes

- None

### Related issues

Fixes https://github.com/hyperlane-xyz/issues/issues/511

### Backward compatibility

Yes, same interface as the previous IGP but for Mailbox V3

### Testing

Unit Tests

### CODE REVIEW

File: rust/Cargo.lock
Comment (yorhodes) [inline rust/Cargo.lock:1]: revert please

File: solidity/contracts/hooks/AbstractMessageIdAuthHook.sol
Comment (yorhodes) [inline solidity/contracts/hooks/AbstractMessageIdAuthHook.sol:1]: assume this is unrelated

File: solidity/contracts/igps/InterchainGasPaymaster.sol [hunk 1]
```diff
@@ -14,13 +17,18 @@ import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/Own
...
     OwnableUpgradeable
 {
+    using Message for bytes;
+    using IGPMetadata for bytes;
     // ============ Constants ============
 
     /// @notice The scale of gas oracle token exchange rates.
     uint256 internal constant TOKEN_EXCHANGE_RATE_SCALE = 1e10;
+    /// @notice default for user call if metadata not provided
+    uint256 internal constant DEFAULT_GAS_USAGE = 69_420;
```
Comment (yorhodes) [inline solidity/contracts/igps/InterchainGasPaymaster.sol:31]: I would make this immutable

File: solidity/contracts/igps/InterchainGasPaymaster.sol [hunk 2]
```diff
@@ -66,36 +74,33 @@ contract InterchainGasPaymaster is
...
+        } else {
+            address refundAddress = metadata.refundAddress();
+            if (refundAddress == address(0))
+                refundAddress = message.senderAddress();
+            payForGas(
+                message.id(),
+                message.destination(),
+                metadata.gasLimit(),
+                refundAddress
+            );
```
Comment (yorhodes) [inline solidity/contracts/igps/InterchainGasPaymaster.sol:102]: can you 
```suggestion
	uint256 gasLimit;
	address refundAddress;
        if (metadata.length == 0) {
            gasLimit = DEFAULT_GAS_USAGE;
            refundAddress = message.senderAddress();
        } else {
            gasLimit = metadata.gasLimit();
            refundAddress = metadata.refundAddress();
        }
        payForGas(
            message.id(),
            message.destination(),
            gasLimit,
            refundAddress
        );
```

File: solidity/contracts/igps/InterchainGasPaymaster.sol [hunk 3]
```diff
@@ -132,6 +137,39 @@ contract InterchainGasPaymaster is
...
+            _destinationDomain,
+            _gasAmount
+        );
+        require(
+            msg.value >= _requiredPayment,
+            "insufficient interchain gas payment"
+        );
+        uint256 _overpayment = msg.value - _requiredPayment;
+        if (_overpayment > 0) {
+            (bool _success, ) = _refundAddress.call{value: _overpayment}("");
```
Comment (yorhodes) [inline solidity/contracts/igps/InterchainGasPaymaster.sol:166]: why use `call` here and `sendValue` elsewhere?

File: solidity/contracts/isms/hook/AbstractMessageIdAuthorizedIsm.sol
Comment (yorhodes) [inline solidity/contracts/isms/hook/AbstractMessageIdAuthorizedIsm.sol:1]: assume this is unrelated

File: solidity/contracts/libs/LibBit.sol
Comment (yorhodes) [inline solidity/contracts/libs/LibBit.sol:1]: ditto

File: solidity/test/igps/InterchainGasPaymaster.t.sol [hunk 4]
```diff
@@ -48,6 +58,66 @@ contract InterchainGasPaymasterTest is Test {
...
+        uint256 _quote = igp.quoteGasPayment(
+            testDestinationDomain,
+            testGasAmount
+        );
+
+        uint256 _overpayment = 25000;
+        bytes memory metadata = abi.encodePacked(
+            uint256(testGasAmount), // gas limit
+            testRefundAddress // refund address
+        );
```
Comment (yorhodes) [inline solidity/test/igps/InterchainGasPaymaster.t.sol:106]: add a `formatMetadata` to the library?

File: solidity/test/isms/OPStackIsm.t.sol
Comment (yorhodes) [inline solidity/test/isms/OPStackIsm.t.sol:1]: unrelated

File: solidity/test/isms/IsmTestUtils.sol
Comment (yorhodes) [inline solidity/test/isms/IsmTestUtils.sol:1]: unrelated

Comment (aroralanuk) [reply solidity/test/isms/IsmTestUtils.sol:1]: related to igp test file

File: solidity/contracts/igps/InterchainGasPaymaster.sol [hunk 5]
```diff
@@ -132,6 +133,38 @@ contract InterchainGasPaymaster is
...
+     * @param _messageId The ID of the message to pay for.
+     * @param _destinationDomain The domain of the message's destination chain.
+     * @param _gasAmount The amount of destination gas to pay for.
+     * @param _refundAddress The address to refund any overpayment to.
+     */
+    function payForGas(
+        bytes32 _messageId,
+        uint32 _destinationDomain,
+        uint256 _gasAmount,
+        address _refundAddress
```
Comment (yorhodes) [inline solidity/contracts/igps/InterchainGasPaymaster.sol:150]: still not sure why this shows up in the diff
in future you should avoid relocating code, especially if it's unchanged to minimize the diff

File: solidity/test/igps/InterchainGasPaymaster.t.sol [hunk 6]
```diff
@@ -48,6 +57,66 @@ contract InterchainGasPaymasterTest is Test {
...
```
Comment (yorhodes) [inline solidity/test/igps/InterchainGasPaymaster.t.sol:105]: use the format helper you added?

**Review by yorhodes — APPROVED** [review]

File: solidity/contracts/igps/InterchainGasPaymaster.sol (hunk 5 above)
Comment (aroralanuk) [reply solidity/contracts/igps/InterchainGasPaymaster.sol:150]: i added postDispatch and git is dumb
